### PR TITLE
instancetype: Introduce Memory.OvercommitPercent

### PR DESCRIPTION
**Pull request by vladikr**
/area instancetype

**What this PR does / why we need it**:
This PR adds the ability to specify the guest RAM overcommit percentage.

So far VM users could create overcommitted guests by manually setting different values for the guest memory and for the virt-launcher pod memory requests.
In order to facilitate this configuration, this PR introduces a new field to the instanceTypes `Spec.Memory.OvercommitPercent`
The virt-launcher pod will request less physical memory by the OvercommitPercent.

**Release note**:
```release-note
Introduce an ability to set memory overcommit percentage in instanceType spec
```

### CODE REVIEW

**Comment by fossedihelm** [comment]
@vladikr Just to be sure I understand correctly. AFAIU instancetype will be applied to vmi during the vm start. From this point of view this percentage will be applied on top of the current memory request, which can be already overcommitted with the `kv.Spec.DeveloperConfiguration.MemoryOvercommit`. Is it correct? Thanks

**Comment by lyarwood** [comment]
/cc
/hold

I'll look at this in more detail today but ultimately this will need to rebased on the v1beta1 PR for v1.0.0

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:901]: `OvercommitPercent` needs some form of validation in the [instance type validation webhook](https://github.com/kubevirt/kubevirt/blob/e145467d7527c7f6b7e2b94e62d22992c410f15d/pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter.go) right?

File: pkg/instancetype/instancetype.go
Comment (vladikr) [reply pkg/instancetype/instancetype.go:906]: Yes. At this point, let's adjust the requests.

File: pkg/instancetype/instancetype.go
Comment (fabiand) [inline pkg/instancetype/instancetype.go:905]: `value` to be `podRequestedMemory`?

**Review by fabiand — COMMENTED** [review]
+1 on the approach to specify the overcommit ratio as a percentage of the guest memory.

File: pkg/instancetype/instancetype.go
Comment (vladikr) [reply pkg/instancetype/instancetype.go:901]: Added.

File: pkg/instancetype/instancetype.go
Comment (vladikr) [reply pkg/instancetype/instancetype.go]: done

**Comment by vladikr** [comment]
> @vladikr Just to be sure I understand correctly. AFAIU instancetype will be applied to vmi during the vm start. From this point of view this percentage will be applied on top of the current memory request, which can be already overcommitted with the `kv.Spec.DeveloperConfiguration.MemoryOvercommit`. Is it correct? Thanks

@fossedihelm sorry, I missed it. I think that `kv.Spec.DeveloperConfiguration.MemoryOvercommit` doesn't play much of a role. On one hand, it is only set when there are no requests set on the VMI.
Instance types will be processed first and then defaults won't be applied. 
On the other hand, even if it wouldn't the case, instanceTypes has a validation and is being rejected if vmi limits are set. 
So overall, I think it is safe.

File: staging/src/kubevirt.io/api/instancetype/v1alpha2/types.go
Comment (iholder101) [inline staging/src/kubevirt.io/api/instancetype/v1alpha2/types.go:156]: supernit: missing newline

File: pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go
Comment (iholder101) [inline pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go:69]: The tests headers should be swapped

File: pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter.go
Comment (iholder101) [inline pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter.go:81]: How about adding an upper bound here?

`>=100` is obviously wrong. Maybe even `>90` should be disallowed / raise a warning

**Review by iholder101 — COMMENTED** [review]
Thanks @vladikr! Looks great.

See some minor comments below

File: pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go
Comment (0xFelix) [inline pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go:83]: What is this line for? Looks wrong to me.

File: pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go
Comment (0xFelix) [inline pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go:87]: This should assert on the error msg `overCommitPercent and hugepages should not be requested together.`.

**Review by 0xFelix — COMMENTED** [review]
Can you please rebase this work on #9575 and move the API changes from v1alpha2 to v1beta1? Thank you!

**Comment by vladikr** [comment]
> Can you please rebase this work on #9575 and move the API changes from v1alpha2 to v1beta1? Thank you!

Yeah, I'm in the process... it's a bit tricky with all the changes

File: pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go
Comment (vladikr) [reply pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter_test.go:69]: :facepalm:

File: pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter.go
Comment (0xFelix) [inline pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter.go:100]: I think it won't work this way, because `admitInstancetypeOrPreference` can be called on instancetypes and preferences. `webhookutils.GetInstanceTypeFromAdmissionRequest` will fail when called on a preference?

**Comment by iholder101** [comment]
> > What happens if a value of `100` resp. a resource request of `0` is used? Does this even make sense?
> 
> Yes, it does. the pod can be entirely "best effort".

what about 110? :)

File: pkg/instancetype/instancetype_test.go
Comment (fossedihelm) [inline pkg/instancetype/instancetype_test.go:1056]: typo which will solve code-lint job:
```suggestion
			It("should apply memory overcommit correctly to VMI", func() {
```

File: staging/src/kubevirt.io/api/instancetype/v1alpha2/types.go
Comment (fossedihelm) [inline staging/src/kubevirt.io/api/instancetype/v1alpha2/types.go:160]: Missing space, it is present in all the other occurrences.
```suggestion
	// +optional
```

File: staging/src/kubevirt.io/api/instancetype/v1beta1/types.go
Comment (fossedihelm) [inline staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:160]: You can add:
```
// +kubebuilder:validation:Maximum=100
// +kubebuilder:validation:Minimum=0
```
and save some manual validation checks.

File: tests/instancetype_test.go
Comment (fossedihelm) [inline tests/instancetype_test.go:209]: kubevirt spec is always restored during global `AfterEach` function for serial tests, hence this can be dropped.

**Review by fossedihelm — COMMENTED** [review]
Some comments and suggestions

File: staging/src/kubevirt.io/api/instancetype/v1beta1/types.go
Comment (vladikr) [reply staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:160]: Thanks!

Comment (iholder101) [reply staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:160]: Wow!! Super cool!

File: pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter.go
Comment (vladikr) [reply pkg/virt-api/webhooks/validating-webhook/admitters/instancetype-admitter.go:100]: Well. For some reason, it worked well for preferences as well, however, I've split the validation for each of these objects to be separate - like any other object we validate.

File: staging/src/kubevirt.io/api/instancetype/v1alpha2/types.go
Comment (0xFelix) [inline staging/src/kubevirt.io/api/instancetype/v1alpha2/types.go:23]: Can you drop this change? Let's clean it up later.

File: pkg/instancetype/instancetype.go
Comment (vladikr) [reply pkg/instancetype/instancetype.go:859]: `Magic number: 100` - what can I say? I'm not ready to introduce const for the number 100 just to make code-lint happy. 
Any other ideas?
I wonder why doesn't it complain about 1

Comment (0xFelix) [reply pkg/instancetype/instancetype.go:859]: 1 must be less magic than 100. :) The test is not required though so we might ignore it.

File: pkg/util/webhooks/webhooks.go
Comment (0xFelix) [inline pkg/util/webhooks/webhooks.go:193]: What about using `request.Object.GetObjectKind()` or `request.RequestKind` to distinguish between `VirtualMachineInstancetype` and `VirtualMachineClusterInstancetype`? It works like this currently because the CRDs are practically identical. But I wonder if this is the cleanest way to do it?

Comment (vladikr) [reply pkg/util/webhooks/webhooks.go:193]: I'm not sure... 
What I mean is that the initial object has almost no importance as long as the `.Spec` is identical in both objects and returns `VirtualMachineInstancetypeSpec`. 
The better approach is that each object will have it's own separate Admit() and validation. But right know, these two objects are being treated similarly. It will make sense to change the object in this function when there will be a distinct difference between the objects.
If you insist I'll add a type switch, but I don't see a reason for this right now.

Comment (0xFelix) [reply pkg/util/webhooks/webhooks.go:193]: Ok, let's leave it like that for now.

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Review by xpivarc — COMMENTED** [review]
/approve

**Comment by lyarwood** [comment]
> > /hold
> > I'll look at this in more detail today but ultimately this will need to rebased on the v1beta1 PR for v1.0.0
> 
> @lyarwood Is there a reason to keep holding the PR?

/unhold

Nope apologies just back from PTO, at the time I just wanted this rebased on v1beta1. Needs another rebase now but otherwise this lgtm!

**Comment by xpivarc** [comment]
@vladikr This needs quick rebase with `make generate`

File: pkg/instancetype/instancetype.go
Comment (fossedihelm) [reply pkg/instancetype/instancetype.go:920]: I just realize that you can add `gomnd` in the `//nolint:dupl,lll,gocyclo` list at the beginning of the file to ignore it.
Basically resulting in:

```
//nolint:dupl,lll,gocyclo,gomnd
package instancetype
[...]
```

**Pull request merged**
